photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java

```diff
@@ -0,0 +1,370 @@
+                    pose = targetPose.transformBy(target.getBestCameraToTarget().inverse());
+                    mili = p.getFirst().getLatestResult().getLatencyMillis();
+                }
+            }
+        }
+        return Pair.of(pose, mili);
+    }
+
+    /** Return the average of the best target poses using ambiguity as weight */
+    private Pair<Pose3d, Double> averageBestTargetsStrategy() {
+        //                  Pair of Double, Double = Ambiguity, Mili
+        List<Pair<Pose3d, Pair<Double, Double>>> tempPoses = new ArrayList<>();
+        double totalAmbiguity = 0;
+        for (int i = 0; i < cameras.size(); i++) {
+            Pair<PhotonCamera, Transform3d> p = cameras.get(i);
+            List<PhotonTrackedTarget> targets = p.getFirst().getLatestResult().targets;
+            for (int j = 0; j < targets.size(); j++) {
+                PhotonTrackedTarget target = targets.get(j);
+                // If the map doesn't contain the ID fail
+                if (!aprilTags.containsKey(target.getFiducialId())) {
+                    DriverStation.reportWarning(
+                            "[RobotPoseEstimator] Tried to get pose of unknown April Tag: "
+                                    + target.getFiducialId(),
+                            false);
+                    continue;
+                }
+                Pose3d targetPose = aprilTags.get(target.getFiducialId());
+                try {
+                    totalAmbiguity += 1. / target.getPoseAmbiguity();
+                } catch (ArithmeticException e) {
+                    DriverStation.reportWarning(
+                            "[RobotPoseEstimator] A total ambiguity of zero exists, using that pose instead!",
+                            false);
+                    return Pair.of(
+                            targetPose.transformBy(target.getBestCameraToTarget().inverse()),
+                            p.getFirst().getLatestResult().getLatencyMillis());
+                }
+                tempPoses.add(
+                        Pair.of(
+                                targetPose.transformBy(target.getBestCameraToTarget().inverse()),
+                                Pair.of(
+                                        target.getPoseAmbiguity(), p.getFirst().getLatestResult().getLatencyMillis())));
+            }
+        }
+
+        Translation3d transform = new Translation3d();
+        Rotation3d rotation = new Rotation3d();
+        double latency = 0;
+
+        for (Pair<Pose3d, Pair<Double, Double>> pair : tempPoses) {
+            try {
+                double weight = (1. / pair.getSecond().getFirst()) / totalAmbiguity;
+                transform = transform.plus(pair.getFirst().getTranslation().times(weight));
+                rotation = rotation.plus(pair.getFirst().getRotation().times(weight));
+                latency += pair.getSecond().getSecond() * weight; // NOTE: Average latency may not work well
```

Comment:
This seems suspect, and no mathematical justification was given in the code. Where are these poses even from?

Reply:
@calcmogul 
Less ambiguous poses need to be weighted more. Poses are from Photon's detection

X_n * (1 / w_n) / (\sigma 1/w_n)
Proof: 
(\sigma 1 / w_n) / (\sigma 1/w_n) = 1

Reply:
But why average them all together? The correct thing to do is give them individually to the pose estimator with different measurement standard deviations. Averaging them into a single pose throws away information.

Reply:
The other problem though is that averaging ambiguous poses is necessarily going to pull your state estimate away from the true value when one of the poses is correct and the other is completely incorrect. The ambiguity metric isn't an accurate reflection of the true probability distribution.

Reply:
Yep, you're 100% right; I went in assuming that the ambiguity would be a true (gaussian) probability distribution -- even then, the latency wouldn't be mathematically correct. 
Any ideas on how to change it? I'm not even sure if it makes since to have this strategy in the first place...

Reply:
> the latency wouldn't be mathematically correct

If the latencies are different, you have to feed them individually to the pose estimator so it can match them up with prior pose estimates before fusing them.

> Any ideas on how to change it? I'm not even sure if it makes since to have this strategy in the first place...

Particle filters and factor graphs can accurately represent multimodal distributions. Those are a lot more work to implement though. I would suggest dropping this strategy since the others handle the distribution a bit better (i.e., pick one target or the other based on some metric, but nothing in between).

Reply:
It's possible to resolve ambiguity if multiple tags are in view by feeding all visible corners to SolvePNP (instead of each tag separately, then fusing each estimate in). This will converge to the true pose.

Reply:
^ I think libapriltag uses some kind of homography thing instead of PnP, but I like that idea.